finished v1 of the address parser feature

smcalilly opened the pull request:
## Overview
This PR is for Sam McAlilly's code challenge submission. The code has an API endpoint which parses address strings and returns the component parts of the address string. The feature includes a UI with a submission form where users can request the address components for specific addresses.

### Demo
Follow the installation directions outlined [in the repo's readme](https://github.com/smcalilly/code-challenge#installation).

You can visit the app at http://localhost:8000. 

## Testing Instructions
#### Manual Testing
- Test the search form with an address string: `123 main st chicago il`. This is the happy path.
- Test the search form with an address string: `123 main st chicago il 123 main st`. This should return an error.
- Test the API endpoint without the UI like this: `http://localhost:8000/api/parse/?address=123%20main%20st%20chicago%20il`. That is the happy path.

#### Automated Testing
In a terminal session in the repo's root directory, run:
```
docker-compose -f docker-compose.yml -f tests/docker-compose.yml run --rm app
```

smcalilly commented on parserator_web/views.py:
```diff
@@ -14,11 +14,25 @@ class AddressParse(APIView):
+            }
+
+            return parsed_address
+        except usaddress.RepeatedLabelError:
```
This error bubbles up to line 23 in `return Response`. If the exception occurs, the error `f"Unable to process address: {address}"` is returned as an API response. After seeing this code in the context of a PR, I think that could be confusing because it's not explicit.

smcalilly replied:
Yeah, I think that moving the error handling into my `get()` is the right approach. That also made me realize that I'd rather be formatting the response dict in the `get()` method, since that method is the request handler. To me, it makes the most sense to prepare the response data there instead of the `parse` method. 

If that doesn't make sense, hopefully you'll see what I'm talking about in the new commit: d0ab7d0

beamalsky replied:
I agree! `parse()` now sticks more closely to what it says it's doing.

beamalsky reviewed (changes requested):
Hey @smcalilly! I'm Bea, and I'll be reviewing your code challenge and co-interviewing you with Derek this week. 

Lots of good stuff here. You demonstrate solid knowledge of Django and ES6 JS conventions, and we found your code clear and well-modularized. I've left some comments inline with questions and places I see some room for refactoring and edge case/error handling.

We'd like you to respond to this review and make the requested changes before your interview. For guidance, we like and try to stick to [Thoughtbot's guide to code review](https://github.com/thoughtbot/guides/tree/master/code-review) when working together on pull requests.

Let me know in the relevant thread if you have any questions, etc., and I'll be sure to get back to you ASAP. Look forward to talking more!

smcalilly commented on parserator_web/views.py:
```diff
@@ -14,11 +14,25 @@ class AddressParse(APIView):
-        return Response({})
+        address = request.query_params.get('address')
+
+        if not address:
```
Ah, yes. Now I'm wondering if I should remove the whitespace on the server side or the client side. I kind of want to do both just to be thorough, what do you think?

beamalsky replied:
I think client side should suffice! Do you see a scenario where you'd want to preserve those extra spaces on a submission?

smcalilly replied:
For the scope of this code challenge's client/server app, all the requests would go through this app's client code. In the future, I could see this API being consumed via other clients which aren't controlled by this code (like if somebody were requesting the API in their own app). So for now, I think client side validation will suffice. If there were ever a task to make the API consumable for other non-DataMade clients, it would be easy to modify the server to trim the whitespace.

However! Since it would be so easy and humans are sometimes unfortunately prone to forget things, I decided to go ahead and trim the whitespace on the server, too. Here is the commit: 0f4bbd2

And regarding "a scenario where you'd want to preserve those extra spaces" — I can't think of a scenario where we'd want to preserve the leading or trailing spaces in an address string. I might be wrong, so let me know if you know of any scenarios that I'm missing.

beamalsky replied:
Very true, and good thinking re: the possible extension of this! One small thing: instead of having to run `input.trim()` twice and also have to check for the existence of the `input` value, I'd use a [conditional operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator) do something like this:

```javascript
	const inputTrimmed = input ? input.trim() : ''

	if (inputTrimmed.length() > 0) {
		const urlWithParams = `${window.location.href}api/parse/?address=${encodeURIComponent(inputTrimmed)}`;
		...
	}
```

smcalilly replied:
Oh cool, I like that. Thanks. Here is the change: a9b5587.

smcalilly commented:
@beamalsky I'm making your suggested changes. They were helpful, thanks. I'm trying to follow the Thoughtbot guide and I'm slightly confused, specifically where it says:

1) Extract some changes and refactoring into future tickets/stories.
2) Link to the code review from the ticket/story. ("Ready for review: https://github.com/organization/project/pull/1")
3) Push commits based on earlier rounds of feedback as isolated commits to the branch. Do not squash until the branch is 
 ready to merge. Reviewers should be able to read individual updates based on their earlier feedback.

Do numbers 1 and 2 apply here? Or can I just follow number 3? That's what I did for my first code change, where I responded to your code review and linked to the commit...is that the correct way of doing this?

smcalilly commented on parserator_web/templates/parserator_web/index.html:
```diff
@@ -39,7 +38,81 @@ <h4>Parsing results</h4>
-    /* TODO: Flesh this out to connect the form to the API and render results
-    in the #address-results div. */
+  (function() {
+    async function getAndRenderAddressComponentsTable(input) {
```
Thanks, though I just refactored it because it wasn't being used like I thought it was. I'm pretty sure that was leftover when I was initially using the JS `fetch` function for the request, but then changed it to an `XMLHttpRequest` since [fetch isn't compatible on some browsers](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#Browser_compatibility).

smcalilly replied:
Dang, I just looked up the browser compatibility of some other stuff I'm doing and it's not all supported.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries#Browser_compatibility
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#Browser_compatibility

beamalsky replied:
Very important point, and good catch! This is something we've gone back and forth about a lot internally at DataMade, and we've decided that the approach we like best here is to write ES6 and then later compile it to vanilla Javascript for browser compatibility. You can see an outline of our exact strategy for what that looks like in Django projects here: https://github.com/datamade/how-to/blob/master/django/django-compressor.md

So rather than writing Javascript explicitly for backwards compatibility, we handle this on a project level and use `django-compressor` as a dependency. That's somewhat outside the scope of this code challenge, so don't worry about implementing that in this PR. This is part of our [`new-django-app` cookiecutter template](https://github.com/datamade/how-to/tree/master/docker/templates), so we have a quick way of easily setting it up as part of new projects.

beamalsky commented:
Good questions. ThoughtBot doesn't use exactly the same stack as us, so there are parts of this that are more about the spirit of generosity and learning together than the mechanics. To answer these points in order:

1. This one doesn't apply to a code challenge. If this were a larger project, and an issue came up that should be dealt with but not in this PR, we'd be opening new Github issues to track those.
2. This one also doesn't apply here. In our usual projects we include [this PR template](https://github.com/datamade/how-to/blob/master/docker/templates/new-django-app/%7B%7Bcookiecutter.app_name%7D%7D/.github/PULL_REQUEST_TEMPLATE.md) to prompt everyone to write up a thorough description of what's going on, and it includes a place ("Closes #XXX") to link back to the relevant Github issue.
3. This is good to follow! Small commits where you can leave a specific message about what you're doing are preferred, and makes review and future documentation easier.

Let me know if any of that still isn't clear!

smcalilly commented on tests/test_views.py:
```diff
@@ -1,15 +1,31 @@
+    response = client.get(url)
+    assert response.status_code == 400
+
+    empty_string = ''
```
Yeah that's cleaner. I included that in this commit: 0f4bbd2

smcalilly commented on parserator_web/templates/parserator_web/index.html:
```diff
@@ -9,26 +9,25 @@
         <h4>Parsing results</h4>
         <p>Address type: <strong><span id="parse-type"></span></strong></p>
-        <table class="table table-bordered">
+        <table class="table table-bordered" id="table">
```
Good point, thanks. Fixed here: 2cf99b5

smcalilly commented on parserator_web/templates/parserator_web/index.html:
```diff
@@ -39,7 +38,81 @@ <h4>Parsing results</h4>
+      p.innerHTML = error;
+      div.appendChild(p);
+
+      setTimeout(() => p.remove(), 3000);
```
The code challenge instructions didn't have any direction on how to handle errors, from a user experience perspective. So I added the timeout because it was quickest/easiest way to clear the error message from the DOM. I almost added a comment in my code about this—I figured this would be a good place to have a discussion about the best way to handle this. Luckily you brought it up without me leaving the comment.

In hindsight, I should've originally done what I just did in this commit: 6033146. That was as simple to do as the timeout. I initially did this when I was first doing the code challenge, but I decided not to because I started to completely refactor the code into something that's more maintainable. 

I think this new commit is clear and maintainable, but it could lead to some unmaintainable code, particularly with managing what's supposed to be on the DOM. If I were to add anymore features to the UI, I'd probably want to refactor it a little bit.  

Explaining by code helps me. Here's an example from my new commit (6033146). I have these two functions that either render the error or remove it. They both reference the `error` class name.
```
function renderErrorMessage(error = 'Oh no! Something bad happened.') {
      const div = document.getElementById('parser');
      const p = document.createElement('p');

      p.className = 'error';
      p.innerHTML = error;
      div.appendChild(p);
    }

function removeErrorMessage() {
      document.querySelectorAll('.error').forEach(element => element.remove());
 }
```

I'm sort of repeating myself. I could create a variable for the error class name, like this:
```
const errorMessageClassName = 'error';
```

That would give me a variable that's easy to change in one place if needed. That essentially creates a reference to an element on the DOM and the basis of an abstraction to do more stuff with the error message. So if I'm gonna do that for one DOM element, then why wouldn't I do that to all of the other elements? I could, but I'd need to refactor some. 

Anyway, that was my thinking. Maybe I'm overthinking it.

smcalilly replied:
Also, I did one more commit on this piece of code: 5275536

beamalsky replied:
I think this is a great solution, and appreciate that you have a modular `clearErrorMessage()` function. Breaking that class name out as a variable wouldn't hurt, but `'error'` is pretty standardized here and more concise so I have no problem leaving that string in.

smcalilly commented:
@beamalsky I've addressed all of your comments. I appreciate your help with making the code better. Let me know if there's anything else I need to do.

beamalsky commented:
Thanks for going through all these comments, Sam! See you tomorrow.
